mypy/fscache.py

```diff
@@ -198,9 +201,21 @@ def isfile_case(self, path: str) -> bool:
         else:
             try:
                 names = self.listdir(head)
+                # This allows to check file name case sensitively in
+                # case-insensitive filesystems.
                 res = tail in names and self.isfile(path)
             except OSError:
                 res = False
+
+        # Also check the other path components in case sensitive way.
+        head, dir = os.path.split(head)
+        while res and head and dir and prefix in head:
```

Comment:
Why `prefix in head`? That does a substring check and I can think of so many edge cases where this goes wrong (e.g. if prefix is `'/'` or `''`). I would implement a more robust check that checks `prefix` is actually a filesystem prefix of `head`. Or you could remove `prefix` from `head` before you even enter the loop. Or you could use `os.path.relpath()`. Or the caller could be requested to pass the prefix and whatever follows separately (so you'd have pass `os.path.join(prefix, head)` to `listdir()`. Or...

Reply:
Indeed.

Reply:
Actually after playing more with these, it looks like `head.startswith(prefix)` is the only one that behaves correctly, while other options I tried fail in some corner cases. Essentially, if the prefix is `''` or `'/'` we should check the whole path in case sensitive manner.

Also could you please show a tricky test case, so that I can add it to the tests?